contrib/babelfishpg_tsql/src/hooks.c

```diff
@@ -5232,39 +5258,147 @@ transform_pivot_clause(ParseState *pstate, SelectStmt *stmt)
+		pivot_context_list = list_make2(list_make1(makeString("bbf_pivot_func")),
+										list_make2((Node *) copyObject(s_sql),
+													(Node *) copyObject(c_sql)
+													)
+								);
+
+		src_sql_string = pstrdup(pstate->p_sourcetext);
+		cat_sql_string = pstrdup(pstate->p_sourcetext);
+
+	}
 
+	in_view = makeNode(A_Const);
+	in_view->val.boolval.type = T_Boolean;
+	in_view->val.boolval.boolval = is_view_transform;
+	in_view->location = -1;
 	/* Store pivot information in FuncCall to live through parser analyzer */
-	pivot_func = makeFuncCall(list_make2(makeString("sys"), makeString("bbf_pivot")), NIL, COERCE_EXPLICIT_CALL, -1);
+	pivot_func = makeFuncCall(list_make2(makeString("sys"), makeString("bbf_pivot")), 
+							  list_make4((Node *) make_string_aconst(src_sql_string, -1),
+										 (Node *) make_string_aconst(cat_sql_string, -1),
+										 (Node *) make_string_aconst(pstrdup(funcName->sval), -1),
+										 (Node *) in_view),
+							  COERCE_EXPLICIT_CALL, 
+							  -1);
 	pivot_func->context = (Node *) pivot_context_list;
 	wrapperSelect_RangeFunction->functions = list_make1(list_make2((Node *) pivot_func, NIL));
 }
 
 static void
 pass_pivot_data_to_fcinfo(FunctionCallInfo fcinfo, Expr *expr)
 {
+	Oid 			func_arg_typeOid;
+	HeapTuple		tuple;
+	char 			*func_name;
+	bool			in_view;
+
 	/* if current FuncExpr is a bbf_pivot function, we set the fcinfo context to pivot data */
 	if (sql_dialect != SQL_DIALECT_TSQL)
 		return;
 
-	if (IsA(expr, FuncExpr) 
-		&& ((FuncExpr*) expr)->context != NULL
-		&& (IsA(((FuncExpr*) expr)->context, List)))
+	func_name = NULL;
+	func_arg_typeOid = InvalidOid;
+
+	/* check if the current function call is bbf_pivot */
+	/* bbf_pivot function expect to pass 4 argument */
+	if (fcinfo->nargs != 4)
+		return;
+
+	tuple = SearchSysCache1(PROCOID, ObjectIdGetDatum(fcinfo->flinfo->fn_oid));
+	if (HeapTupleIsValid(tuple))
+	{
+		Form_pg_proc form = (Form_pg_proc) GETSTRUCT(tuple);
+		func_name = NameStr(form->proname);
+		func_arg_typeOid = form->proargtypes.values[0];
+		ReleaseSysCache(tuple);
+	}
+
+	if (func_name == NULL || strcmp(func_name, "bbf_pivot") != 0 || func_arg_typeOid != TEXTOID)
+		return;
+
+	/* At this point, we are certain the function is bbf_pivot*/
+	in_view = PG_GETARG_BOOL(3);
+
+	if (in_view)
 	{
-		Node *node;	
-		node = list_nth((List *)((FuncExpr*) expr)->context, 0);
-		if (IsA(node, List) 
-				&& IsA(list_nth((List *)node, 0), String) 
-				&& strcmp(((String *)list_nth((List *)node, 0))->sval, "bbf_pivot_func") == 0)
+		char	*src_sql_string;
+		char	*cat_sql_string;
+		List 	*pivot_context_list;
+		Node 	*restored_src_rawstmt;
+		Node 	*restored_cat_rawstmt;
+
+		src_sql_string =  text_to_cstring(PG_GETARG_TEXT_PP(0));
+		cat_sql_string =  text_to_cstring(PG_GETARG_TEXT_PP(1));
+
+		restored_src_rawstmt = linitial(raw_parser(src_sql_string, RAW_PARSE_DEFAULT));
+		restored_cat_rawstmt = linitial(raw_parser(cat_sql_string, RAW_PARSE_DEFAULT));
```

Comment:
we don't have a parsing error here probably because we skipped the ANTLR parser. To make it more protected, I will guard those two with set sql dialect to psql and then tsql.